File: day5/5.py
from timeit import default_timer as timer

def get_input():
    with open("./day5/5.in","r") as f:
        return [line.strip()for line in f.readlines()]


p1start = timer()
res = []
decoding = {'F':'0', 'B':'1', 'L':'0', 'R':'1'}
for barcode in get_input():
    row, col = barcode[:7], barcode[7:]
    row_bin, col_bin = '0b', '0b'
    for letter in row:
        row_bin += decoding[letter]
    for letter in col:
        col_bin += decoding[letter]
    row, col = int(row_bin, 2), int(col_bin, 2)
    res.append((row*8) + col)
p1 = max(res)
p1end = timer()

p2start = timer()
res.sort()
for i in range(len(res)):
    if res[i+1] - res[i] == 2:
        p2 = res[i] + 1
        break
p2end = timer()


print("Day 5: Binary Boarding")
print(f"Part 1: {p1} in {p1end-p1start}s.")
print(f"Part 2: {p2} in {p2end-p2start}s.")

File: day5/5.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"time"

	"../helpers"
)

func p1(data []string) []int {
	defer helpers.TimeTrack(time.Now())
	res := make([]int, 1)
	decoding := map[rune]string{'F': "0", 'B': "1", 'L': "0", 'R': "1"}
	for _, barcode := range data {
		row, col := barcode[:7], barcode[7:]
		rowBin, colBin := "", ""
		for _, letter := range row {
			rowBin = rowBin + decoding[letter]
		}
		for _, letter := range col {
			colBin = colBin + decoding[letter]
		}
		rowNum, _ := strconv.ParseInt(rowBin, 2, 64)
		colNum, _ := strconv.ParseInt(colBin, 2, 64)
		res = append(res, (int(rowNum)*8)+int(colNum))
	}
	_, max := helpers.MinMax(res)
	fmt.Printf("Part 1: %d ", max)
	return res
}

func p2(data []string, ids []int) {
	defer helpers.TimeTrack(time.Now())
	var res int
	sort.Ints(ids)
	for i := 0; i < len(ids); i++ {
		if ids[i+1]-ids[i] == 2 {
			res = ids[i] + 1
			break
		}
	}
	fmt.Printf("Part 2: %d ", res)
}

func main() {
	fmt.Println("Day 5: Binary Boarding")
	fileName := "/Users/jack/Documents/Github/aoc-2020/day5/5.in"
	data, err := helpers.GetInputStringByLine(fileName)
	if err != nil {
		panic(err)
	}
	res := p1(data)
	p2(data, res)
}
